server/src/main/java/org/elasticsearch/index/codec/vectors/reflect/OffHeapReflectionUtils.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.index.codec.vectors.reflect;
+
+import org.apache.lucene.codecs.hnsw.FlatVectorsReader;
+import org.apache.lucene.codecs.lucene99.Lucene99FlatVectorsReader;
+import org.apache.lucene.codecs.lucene99.Lucene99HnswVectorsReader;
+import org.apache.lucene.codecs.lucene99.Lucene99ScalarQuantizedVectorsReader;
+import org.apache.lucene.index.FieldInfo;
+import org.apache.lucene.index.VectorEncoding;
+import org.elasticsearch.core.SuppressForbidden;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.VarHandle;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.util.Map;
+
+import static java.lang.invoke.MethodType.methodType;
+
+/**
+ * Reflective access to non-accessible members of Lucene's KnnVectorsReader implementations.
+ * Remove once KnnVectorsReaders::getOffHeapByteSize is available.
+ */
+public class OffHeapReflectionUtils {
+
+    private OffHeapReflectionUtils() {}
+
+    static final String FLAT_VECTOR_DATA_EXTENSION = "vec";
+    static final String SQ_VECTOR_INDEX_EXTENSION = "veq";
+    static final String HNSW_VECTOR_INDEX_EXTENSION = "vex";
+
+    private static final MethodHandle GET_FIELD_ENTRY_HNDL_SQ;
+    private static final MethodHandle GET_VECTOR_DATA_LENGTH_HANDLE_SQ;
+    private static final VarHandle RAW_VECTORS_READER_HNDL_SQ;
+    private static final MethodHandle GET_FIELD_ENTRY_HANDLE_L99FLT;
+    private static final MethodHandle VECTOR_DATA_LENGTH_HANDLE_L99FLT;
+    private static final MethodHandle GET_FIELD_ENTRY_HANDLE_L99HNSW;
+    private static final MethodHandle GET_VECTOR_INDEX_LENGTH_HANDLE_L99HNSW;
+    private static final VarHandle FLAT_VECTORS_READER_HNDL_L99HNSW;
+
+    static final Class<?> L99_SQ_VR_CLS = Lucene99ScalarQuantizedVectorsReader.class;
+    static final Class<?> L99_FLT_VR_CLS = Lucene99FlatVectorsReader.class;
+    static final Class<?> L99_HNSW_VR_CLS = Lucene99HnswVectorsReader.class;
+
+    static {
+        try {
+            // Lucene99ScalarQuantizedVectorsReader
+            var cls = Class.forName("org.apache.lucene.codecs.lucene99.Lucene99ScalarQuantizedVectorsReader$FieldEntry");
```

Comment:
I realize it would be even more complicated, but there are all the bwc indices from Lucene (Lucene95, 92, etc.). Do we want to include those? Its probable that folks have data in those old indices. 

Its likely fine to do a "good enough" job here and indicate that stats are only available on newer formats.